debugger/Main.hs

```diff
@@ -21,19 +21,126 @@ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+          ]
+
+addOption :: Config -> TOption -> IO Config
+addOption config (DebugDumpFile f) = return config { confDebugDumpFile = f}
+addOption config (Datalog f)       = return config { confDatalogFile  = f}
+addOption config (LibDir d)        = return config { confLibDirs = nub (d:confLibDirs config)}
+addOption config (OutputDir d)     = return config { confOutputDir = d }
+addOption config Help              = return config { confAction = ActionHelp}
+addOption config OutputInternal   = return config { confOutputInternal = True }
+addOption config (OutputInput p)  = return config { confOutputInput = p }
+addOption config DumpFlat         = return config { confDumpFlat = True }
+
+validateConfig :: Config -> IO ()
+validateConfig Config{..} = do
+    when (confDatalogFile == "" && confAction /= ActionHelp && confAction /= ActionVersion)
+         $ errorWithoutStackTrace "input file not specified"
 
 main :: IO ()
 main = do
-       (filename:_) <- getArgs
-       contents <- readFile filename
-       let result = parseText contents
-       putStr result
+       args <- getArgs
+       prog <- getProgName
+       home <- lookupEnv "DDLOG_HOME"
+       config <- case getOpt Permute options args of
+                      (flags, [], []) -> do
+                                            conf <- foldM addOption defaultConfig flags
+                                            validateConfig conf
+                                            return conf
+                                         `catch`
+                                         (\e -> do putStrLn $ usageInfo ("Usage: " ++ prog ++ " [OPTION...]") options
+                                                   throw (e::SomeException))
+                      _ -> errorWithoutStackTrace $ usageInfo ("Usage: " ++ prog ++ " [OPTION...]") options
+       config' <- case home of
+           Just(p) -> addOption config (LibDir $ p ++ "/lib")
+           _       -> return config
+       do
+            datalogProg <- parseProgram config'
+            events <- parseEventFromDumpFile config'
+            let recordMap = handleDebugEvents events emptyDebuggerMaps datalogProg
+                s = queryAll events (dbgRecordMap recordMap)
+            putStr $ show s
+
+queryAll :: [Event] -> DebuggerRecordMap -> String
+queryAll [] _ = "Finished"
+queryAll (event: events) dgbRecordMap =
+    let outputRecord = evtOutput event
+        operatorId = evtOperatorId event
+        operatorInput = InputOp operatorId
+        dbgRecord = DebuggerRecord {dbgRecord=outputRecord, dbgOperatorId=operatorInput}
+        dbgRecordNodes = queryDerivations dbgRecord dgbRecordMap
+    in trace (show dbgRecordNodes) $ queryAll events dgbRecordMap
+
+parseProgram :: Config -> IO (DatalogProgram)
+parseProgram Config{..} = do
+    fdata <- readFile confDatalogFile
+    (d, _, _) <- parseDatalogProgram (takeDirectory confDatalogFile:confLibDirs) True fdata confDatalogFile
+    d'' <- case confOutputInternal of
```

Comment:
This means that we have to ensure that we pass the same options when we compiled the ddlog program and  here.
Instead of loading the original program here (and ensuring the same options are passed), I wonder if we can just load the dumped transformed source here? I haven't thought about it deeply yet, so not sure if it's possible to generate DatalogProgram from an ast file?